### PR TITLE
chore: Enable govet linter and fix reported errors

### PR DESCRIPTION
**Pull request by ghislainbourgeois**
Enables the `govet` linter and fix reported errors.

### CODE REVIEW

**Comment by gab-arrobo** [comment]
ok to test

**Comment by gab-arrobo** [comment]
@ghislainbourgeois, FYI. `This branch is out-of-date with the base branch`

**Comment by ghislainbourgeois** [comment]
> @ghislainbourgeois, FYI. `This branch is out-of-date with the base branch`

Fixed

File: gmm/handler.go [hunk 1]
```diff
@@ -1622,13 +1623,14 @@ func HandleUeSliceInfoDelete(ue *context.AmfUe, accessType models.AccessType, ns
 
 	// we are doing local cleanup for now, below code will be deprecated when we support
 	// Configuration Update
+	var problemDetail *models.ProblemDetails
```
Comment (gab-arrobo) [inline gmm/handler.go:1626]: To be aligned to the previous change, should not this be `problemDetails`?
```suggestion
	var problemDetails *models.ProblemDetails
```

File: gmm/handler.go [hunk 2]
```diff
@@ -1622,13 +1623,14 @@ func HandleUeSliceInfoDelete(ue *context.AmfUe, accessType models.AccessType, ns
...
 	ue.SmContextList.Range(func(key, value interface{}) bool {
 		smContext := value.(*context.SmContext)
 		if smContext.Snssai().Sst == int32(nssai.Sst) && smContext.Snssai().Sd == nssai.Sd {
 			logger.GmmLog.Infof("Deleted Slice [sst: %v, sd: %v]matched with smcontext, sending Release SMContext Request to SMF",
 				smContext.Snssai().Sst, smContext.Snssai().Sd)
 			// send smcontext release request
-			problemDetail, err := consumer.SendReleaseSmContextRequest(ue, smContext, nil, "", nil)
+			problemDetail, err = consumer.SendReleaseSmContextRequest(ue, smContext, nil, "", nil)
 			if problemDetail != nil {
 				ue.GmmLog.Errorf("Release SmContext Failed Problem[%+v]", problemDetail)
```
Comment (gab-arrobo) [inline gmm/handler.go:1635]: In alignment to my previous comment
```suggestion
			problemDetails, err = consumer.SendReleaseSmContextRequest(ue, smContext, nil, "", nil)
			if problemDetails != nil {
				ue.GmmLog.Errorf("Release SmContext Failed Problem[%+v]", problemDetails)
```

File: context/amf_ue.go [hunk 3]
```diff
@@ -69,7 +69,7 @@ type AmfUe struct {
 	// Mutex sync.Mutex `json:"mutex,omitempty" yaml:"mutex" bson:"mutex,omitempty"`
 	Mutex sync.Mutex `json:"-"`
 	/* the AMF which serving this AmfUe now */
-	servingAMF *AMFContext `json:"servingAMF,omitempty"` // never nil
+	servingAMF *AMFContext // never nil
```
Comment (gab-arrobo) [inline context/amf_ue.go:72]: Why not doing the following (with the corresponding changes in the respective files)?
```suggestion
	ServingAMF *AMFContext `json:"servingAMF,omitempty"` // never nil
```

File: context/amf_ue.go [hunk 4]
```diff
@@ -147,9 +147,9 @@ type AmfUe struct {
 	SmContextList sync.Map `json:"-"` // map[int32]*SmContext, pdu session id as key
 	/* Related Context*/
 	//RanUe map[models.AccessType]*RanUe `json:"ranUe,omitempty" yaml:"ranUe" bson:"ranUe,omitempty"`
-	RanUe map[models.AccessType]*RanUe `json:"ranUe, omitEmpty"`
+	RanUe map[models.AccessType]*RanUe `json:"ranUe,omitEmpty"`
 	/* other */
-	onGoing                       map[models.AccessType]*OnGoing        `json:"onGoing,omitempty"`
+	onGoing                       map[models.AccessType]*OnGoing
```
Comment (gab-arrobo) [inline context/amf_ue.go:152]: Why not doing the following (with the corresponding changes in the respective files)?
```suggestion
	OnGoing                       map[models.AccessType]*OnGoing        `json:"onGoing,omitempty"`
```

File: ngap/ngap_test.go
Comment (gab-arrobo) [inline ngap/ngap_test.go:1]: Why are these changes needed? There changes are not due to enabling `govet`, correct?

Comment (ghislainbourgeois) [reply ngap/ngap_test.go:1]: Yes and no, the changes made in `metrics/kafka.go` made this test fail because of a nil pointer. It was working before because the default value was an empty struct and a method could be dispatched, but the default value of a pointer is nil.

File: context/amf_ue.go (hunk 3 above)
Comment (ghislainbourgeois) [reply context/amf_ue.go:72]: I tried doing it for this particular case, but it unfortunately clashes with a method named exactly like that. I could change the name of the attribute further, or change the method name, but this will be more involved.

File: gmm/handler.go (hunk 1 above)
Comment (ghislainbourgeois) [reply gmm/handler.go]: Done

File: gmm/handler.go (hunk 2 above)
Comment (ghislainbourgeois) [reply gmm/handler.go]: Done

File: context/amf_ue.go (hunk 4 above)
Comment (ghislainbourgeois) [reply context/amf_ue.go:152]: This one also is problematic because a struct and a method already exist with that name. I will try to rename them and see how this goes.

File: context/amf_ue.go (hunk 3 above)
Comment (ghislainbourgeois) [reply context/amf_ue.go]: I pushed a new commit that does this rename and also gets rid of the method for a direct attribute access instead.

**Comment by gab-arrobo** [comment]
@ghislainbourgeois, did you have a chance to run an end-to-end test using the changes from this PR? I am asking this because I am running a test using AiaB (w/ gnbsim) and the test fails as shown below:

```
$ make 5g-test
...
...
2024-01-03T17:29:41Z [INFO][GNBSIM][Profile][profile2] ExecuteProfile ended
2024-01-03T17:29:41Z [INFO][GNBSIM][Summary] Profile Name: profile2 , Profile Type: pdusessest
2024-01-03T17:29:41Z [INFO][GNBSIM][Summary] Ue's Passed: 0 , Ue's Failed: 5
2024-01-03T17:29:41Z [INFO][GNBSIM][Summary] Profile Errors:
2024-01-03T17:29:41Z [ERRO][GNBSIM][Summary] imsi:imsi-208930100007487, profile timeout
2024-01-03T17:29:41Z [ERRO][GNBSIM][Summary] imsi:imsi-208930100007488, profile timeout
2024-01-03T17:29:41Z [ERRO][GNBSIM][Summary] imsi:imsi-208930100007489, profile timeout
2024-01-03T17:29:41Z [ERRO][GNBSIM][Summary] imsi:imsi-208930100007490, profile timeout
2024-01-03T17:29:41Z [ERRO][GNBSIM][Summary] imsi:imsi-208930100007491, profile timeout
2024-01-03T17:29:41Z [INFO][GNBSIM][Summary] Profile Status: FAIL
make: *** [Makefile:679: 5g-test] Error 1
```

**Comment by ghislainbourgeois** [comment]
> @ghislainbourgeois, did you have a chance to run an end-to-end test using the changes from this PR? I am asking this because I am running a test using AiaB (w/ gnbsim) and the test fails as shown below:
> 
> ```
> $ make 5g-test
> ...
> ...
> 2024-01-03T17:29:41Z [INFO][GNBSIM][Profile][profile2] ExecuteProfile ended
> 2024-01-03T17:29:41Z [INFO][GNBSIM][Summary] Profile Name: profile2 , Profile Type: pdusessest
> 2024-01-03T17:29:41Z [INFO][GNBSIM][Summary] Ue's Passed: 0 , Ue's Failed: 5
> 2024-01-03T17:29:41Z [INFO][GNBSIM][Summary] Profile Errors:
> 2024-01-03T17:29:41Z [ERRO][GNBSIM][Summary] imsi:imsi-208930100007487, profile timeout
> 2024-01-03T17:29:41Z [ERRO][GNBSIM][Summary] imsi:imsi-208930100007488, profile timeout
> 2024-01-03T17:29:41Z [ERRO][GNBSIM][Summary] imsi:imsi-208930100007489, profile timeout
> 2024-01-03T17:29:41Z [ERRO][GNBSIM][Summary] imsi:imsi-208930100007490, profile timeout
> 2024-01-03T17:29:41Z [ERRO][GNBSIM][Summary] imsi:imsi-208930100007491, profile timeout
> 2024-01-03T17:29:41Z [INFO][GNBSIM][Summary] Profile Status: FAIL
> make: *** [Makefile:679: 5g-test] Error 1
> ```

I did not test with AiaB, I will investigate.

**Comment by gab-arrobo** [comment]
> 
> I did not test with AiaB, I will investigate.

Well, you do not have to use AiaB, but my point was to check whether you had tested the changes using an end-to-end test (e.g., AiaB, OnRamp, etc.)

**Comment by ghislainbourgeois** [comment]
> > I did not test with AiaB, I will investigate.
> 
> Well, you do not have to use AiaB, but my point was to check whether you had tested the changes using an end-to-end test (e.g., AiaB, OnRamp, etc.)

I had not yet no. I am able to reproduce the issue, and it seems to be because the mutex is uninitialized. I will push a fix.

**Comment by ghislainbourgeois** [comment]
I pushed a fix that I was able to confirm working end-to-end with gnbsim.

**Review by gab-arrobo — APPROVED** [review]
+1 (tested changes with AiaB)
